library/alloc/src/str.rs

```diff
@@ -619,3 +637,60 @@ impl str {
 pub unsafe fn from_boxed_utf8_unchecked(v: Box<[u8]>) -> Box<str> {
     unsafe { Box::from_raw(Box::into_raw(v) as *mut str) }
 }
+
+/// Converts the bytes while the bytes are still ascii.
+/// This is performed in chunks of `16*size_of::<usize>()` which in testing
+/// performed the fastest.
+#[inline]
+#[cfg(not(test))]
+#[cfg(not(no_global_oom_handling))]
+fn convert_while_ascii(b: &[u8], out: &mut Vec<u8>, convert: fn(&u8) -> u8) {
+    out.reserve(b.len());
+
+    const USIZE_SIZE: usize = mem::size_of::<usize>();
+    const MAGIC_UNROLL: usize = 16;
+
+    let mut i = 0;
+    while i + USIZE_SIZE * MAGIC_UNROLL <= b.len() {
+        unsafe {
+            let range = i..i + USIZE_SIZE * MAGIC_UNROLL;
+
+            // Safety: we have checks the sizes `b` and `out` to know that our
+            let in_chunk = b.get_unchecked(range.clone());
+            let out_chunk = out.spare_capacity_mut().get_unchecked_mut(range);
+
+            let mut count = 0;
+            for j in 0..MAGIC_UNROLL {
+                // read the bytes 1 usize at a time (unaligned since we haven't checked the alignment)
+                // safety: in_chunk is valid bytes in the range
+                let chunk = in_chunk.as_ptr().cast::<usize>().add(j).read_unaligned();
+                count += contains_nonascii(chunk) as usize;
```

Comment:
Have you tried bitoring the results of the contains_nonascii masking, and then testing once at the end? Might be better throughput.

Reply:
That's smart, hadn't thought of that. Will try it

Reply:
There was a marginal improvement. Thanks